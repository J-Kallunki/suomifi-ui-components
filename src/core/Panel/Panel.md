```jsx
<Panel className="panel-test">Test</Panel>
<Panel.expansion title="Test expansion" className="panel-expansion-test">Test expansion content</Panel.expansion>
<Panel.expansionGroup OpenAll="Open all" CloseAll="Close all">
  <Panel.expansion title="Test expansion 1">Test expansion content 1</Panel.expansion>
  <Panel.expansion title="Test expansion 2">Test expansion content 2</Panel.expansion>
  <Panel.expansion title="Test expansion 3">Test expansion content 3</Panel.expansion>
</Panel.expansionGroup>
```

### Example with controlled open-prop:

```jsx
<Component initialState={{ open: true }}>
  {({ setState, state }) => (
    <Panel.expansion
      title="Test expansion bgr"
      noPadding
      open={state.open}
      onClick={() => setState({ open: !state.open })}
    >
      <div style={{ backgroundColor: 'green', padding: '20px' }}>
        Test expansion content
      </div>
    </Panel.expansion>
  )}
</Component>
```
