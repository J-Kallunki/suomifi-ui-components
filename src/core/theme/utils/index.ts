export { withDefaultTheme } from './defaultTheme';
export { focus } from './focus';
export { boxshadowOutline } from './outline';
export { margin, padding } from './spacing';
