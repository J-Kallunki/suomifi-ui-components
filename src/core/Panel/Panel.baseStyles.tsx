import { css } from '@emotion/core';
import { suomifiTheme } from '../theme';
import { PanelProps } from './Panel';
import { element, font } from '../theme/reset';

export const baseStyles = ({ theme = suomifiTheme }: PanelProps) => css`
  ${element}
  ${font}
  padding: 20px;
  background-color: ${theme.colors.white};
`;
