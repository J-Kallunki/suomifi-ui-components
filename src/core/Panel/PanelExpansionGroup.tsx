import React, { Component, ReactElement } from 'react';
import styled from '@emotion/styled';
import { withDefaultTheme } from '../theme/utils/defaultTheme';
import { ThemeComponent } from '../theme';
import {
  PanelExpansionGroup as CompPanelExpansionGroup,
  PanelExpansionGroupProps as CompPanelExpansionGroupProps,
} from '../../components/Panel/PanelExpansionGroup';
import { Button, ButtonProps } from '../Button/Button';
import { baseStyles } from './PanelExpansionGroup.baseStyles';

export interface PanelExpansionGroupProps
  extends CompPanelExpansionGroupProps,
    ThemeComponent {}

const StyledPanelExpansionGroup = styled(
  ({ theme, ...passProps }: PanelExpansionGroupProps) => (
    <CompPanelExpansionGroup {...passProps} />
  ),
)`
  ${props => baseStyles(props)};
`;

const OpenAllButton = (children: ReactElement<ButtonProps>) => (
  <Button.unstyled>{children}</Button.unstyled>
);

/**
 * Used for panel style and defined actions
 */
export class PanelExpansionGroup extends Component<PanelExpansionGroupProps> {
  render() {
    const { OpenAll, CloseAll, ...passProps } = withDefaultTheme(this.props);
    return (
      <StyledPanelExpansionGroup
        {...passProps}
        OpenAll={OpenAllButton(OpenAll)}
        CloseAll={OpenAllButton(CloseAll)}
      />
    );
  }
}
