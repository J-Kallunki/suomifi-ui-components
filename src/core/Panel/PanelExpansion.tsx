import React, { Component, Fragment } from 'react';
import styled from '@emotion/styled';
import { defaultPropsTheme } from '../utils';
import { suomifiTheme, ThemeComponent } from '../theme';
import { baseStyles as panelBaseStyles } from './Panel.baseStyles';
import { baseStyles } from './PanelExpansion.baseStyles';
import {
  PanelExpansion as CompPanelExpansion,
  PanelExpansionProps as CompPanelExpansionProps,
} from '../../components/Panel/PanelExpansion';
import { Icon } from '../Icon/Icon';
import classnames from 'classnames';

export interface PanelExpansionProps
  extends CompPanelExpansionProps,
    ThemeComponent {
  /** Remove padding from expandable content area (for background usage with padding in given container etc.) */
  noPadding?: boolean;
}

const StyledPanelExpansion = styled(
  ({ noPadding, ...passProps }: PanelExpansionProps) => {
    return <CompPanelExpansion {...passProps} />;
  },
)`
  ${props => panelBaseStyles(props)};
  ${props => baseStyles(props)};
`;

interface PanelExpansionState {
  openState: boolean;
}

/**
 * Used for panel style and defined actions
 */
export class PanelExpansion extends Component<PanelExpansionProps> {
  static defaultProps = defaultPropsTheme(CompPanelExpansion);

  /** State is only used to update the caret-icon */
  state: PanelExpansionState = {
    openState:
      this.props.defaultOpen !== undefined ? this.props.defaultOpen : false,
  };

  handleClick = () => {
    const { open, onClick } = this.props;
    const { openState } = this.state;
    const notControlled = open === undefined;
    if (notControlled) {
      this.setState({ openState: !openState });
    }
    if (!!onClick) {
      onClick({ openState: notControlled ? !openState : !!open });
    }
  };

  render() {
    // TODO why default theme neede here when should be set by defaultProps??
    const {
      open,
      title,
      titleTag,
      theme = suomifiTheme,
      ...passProps
    } = this.props;
    const notControlled = open === undefined;
    const openState = !notControlled ? open : this.state.openState;
    return (
      <StyledPanelExpansion
        {...passProps}
        theme={theme}
        onClick={this.handleClick}
        open={open}
        titleTag={titleTag}
        title={
          <Fragment>
            {title}
            {!titleTag && (
              <Icon
                icon="chevronDown"
                className={classnames('fi-panel-expansion-title-icon', {
                  'fi-panel-expansion-title-icon--open': openState,
                })}
                color={theme.colors.secondaryColor}
              />
            )}
          </Fragment>
        }
      />
    );
  }
}
