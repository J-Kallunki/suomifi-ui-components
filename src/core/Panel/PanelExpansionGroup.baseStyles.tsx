import { css } from '@emotion/core';
import { suomifiTheme } from '../theme';
import { PanelExpansionGroupProps } from './PanelExpansionGroup';
import { element, fontPanelTitle, focus } from '../theme/reset';

export const baseStyles = ({
  theme = suomifiTheme,
}: PanelExpansionGroupProps) => css`
  ${element}
  display: flex;
  flex-direction: column;
  & > .fi-panel-expansion-group-panels {
    flex: none;

    & .fi-panel-expansion {
      margin-top: 0;
      margin-bottom: 0;
      transition: margin ${theme.transitions.basicTime}
        ${theme.transitions.basicTimingFunction};
      &.fi-panel-expansion--open {
        &:not(:first-child) {
          margin-top: 14px;
        }
        &:not(:last-child) {
          margin-bottom: 14px;
        }
      }
    }
  }

  & > .fi-panel-expansion-group-all-button {
    ${element}
    ${fontPanelTitle}
    ${focus}
    flex: 1;
    margin-left: auto;
    margin-bottom: 10px;
    padding: 4px 0;
    color: ${theme.colors.secondaryColor};
    cursor: pointer;
  }
`;
