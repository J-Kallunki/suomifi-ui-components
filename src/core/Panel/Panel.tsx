import React, { Component } from 'react';
import styled from '@emotion/styled';
import { withDefaultTheme } from '../theme/utils/defaultTheme';
import { ThemeComponent } from '../theme';
import { baseStyles } from './Panel.baseStyles';
import {
  Panel as CompPanel,
  PanelProps as CompPanelProps,
} from '../../components/Panel/Panel';
import { PanelExpansion, PanelExpansionProps } from './PanelExpansion';
import {
  PanelExpansionGroup,
  PanelExpansionGroupProps,
} from './PanelExpansionGroup';

type PanelVariant = 'default' | 'expansion' | 'expansionGroup';

export interface PanelProps extends CompPanelProps, ThemeComponent {
  /**
   * 'default' | 'expansion'
   * @default default
   */
  variant?: PanelVariant;
}

const StyledPanel = styled(({ theme, ...passProps }: PanelProps) => (
  <CompPanel {...passProps} />
))`
  ${props => baseStyles(props)};
`;

type VariantPanelProps =
  | PanelProps
  | PanelExpansionProps & { variant?: PanelVariant }
  | PanelExpansionGroupProps & { variant?: PanelVariant };

/**
 * Used for panel style and defined actions
 */
export class Panel extends Component<VariantPanelProps> {
  static expansion = (props: PanelExpansionProps) => {
    return <PanelExpansion {...props} />;
  };

  static expansionGroup = (props: PanelExpansionGroupProps) => {
    return <PanelExpansionGroup {...props} />;
  };

  render() {
    const { variant, ...passProps } = withDefaultTheme(this.props);
    if (variant === 'expansion') {
      return <PanelExpansion {...passProps as PanelExpansionProps} />;
    }
    if (variant === 'expansionGroup') {
      return <PanelExpansionGroup {...passProps as PanelExpansionGroupProps} />;
    }
    return <StyledPanel {...passProps} />;
  }
}
