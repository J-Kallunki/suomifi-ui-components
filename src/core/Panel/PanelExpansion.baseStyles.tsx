import { css } from '@emotion/core';
import { suomifiTheme } from '../theme';
import { PanelExpansionProps } from './PanelExpansion';
import { element, focus, fontPanelTitle } from '../theme/reset';
import { absolute } from '../../components/utils/pseudo';

export const baseStyles = ({
  theme = suomifiTheme,
  noPadding = false,
}: PanelExpansionProps) => css`
  ${absolute('before')}
  position: relative;
  padding: 0;
  border-radius: 2px;
  box-shadow: ${theme.shadows.panelShadow};

  &:before {
    background-color: ${theme.colors.panelExpansionBgr};
    opacity: 0;
  }

  &.fi-panel-expansion--open:before {
    opacity: 1;
    transition: opacity ${theme.transitions.basicTime}
      ${theme.transitions.basicTimingFunction};
  }

  & .fi-panel-expansion-title {
    ${element}
    ${focus}
    position: relative;
    display: block;
    width: 100%;
    &--no-tag {
      ${fontPanelTitle}
      padding: 20px 60px 20px 20px;
      color: ${theme.colors.secondaryColor};
    }
  }
  & .fi-panel-expansion-title-icon {
    position: absolute;
    top: 0;
    right: 0;
    margin: 20px;
  }
  & .fi-panel-expansion-title--open .fi-panel-expansion-title-icon,
  & .fi-panel-expansion-title-icon--open {
    transform: rotate(-180deg);
  }

  & > .fi-panel-expansion-content {
    position: relative;
    ${!noPadding && 'padding: 0 20px 20px 20px;'}
  }
`;
