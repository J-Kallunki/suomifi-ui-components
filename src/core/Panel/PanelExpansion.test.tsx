// import React from 'react';
// import { render } from 'react-testing-library';

// import { PanelExpansion } from './PanelExpansion';
import { cssFromBaseStyles } from '../utils';
import { baseStyles } from './Panel.baseStyles';

// const Title1 = () => (
//   <span data-testid="panel-expansion-title">Test expansion</span>
// );
// const Title2 = () => (
//   <span data-testid="panel-expansion-title-2">Test expansion two</span>
// );

// test('calling render with the same component on the same container does not remount', () => {
//   const panelRenderer = render(
//     <PanelExpansion
//       title={Title1}
//       titleTag="h3"
//       className="panel-expansion-test"
//       data-testid="panel-expansion"
//     >
//       Test expansion content
//     </PanelExpansion>,
//   );
//   const { getByTestId, container, rerender } = panelRenderer;
//   expect(container.firstChild).toMatchSnapshot();
//   expect(getByTestId('panel-expansion-title').textContent).toBe(
//     'Test expansion',
//   );

//   // re-render the same component with different props
//   rerender(
//     <PanelExpansion title={Title2}>Test expansion content</PanelExpansion>,
//   );
//   expect(getByTestId('panel-expansion-title-2').textContent).toBe(
//     'Test expansion two',
//   );
// });

test('CSS export', () => {
  const css = cssFromBaseStyles(baseStyles);
  expect(css).toEqual(expect.stringContaining('background-color'));
});
