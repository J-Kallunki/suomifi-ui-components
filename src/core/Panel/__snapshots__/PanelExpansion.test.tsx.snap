// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calling render with the same component on the same container does not remount 1`] = `
.emotion-3 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font-size: 100%;
  font: inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: auto;
  display: block;
  max-width: 100%;
  display: block;
  width: 100%;
  max-width: 100%;
  color: #282828;
  font-family: 'Source Sans Pro','Helvetica Neue',Arial;
  font-size: 18px;
  line-height: 18px;
  font-weight: 400;
  line-height: 1em;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  white-space: nowrap;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-font-smoothing: antialiased;
  padding: 20px;
  background-color: #FFFFFF;
  position: relative;
  position: relative;
  padding: 0;
  border-radius: 2px;
  box-shadow: 0 0 4px 0 rgba(40,40,40,0.3);
}

.emotion-3:before {
  content: '';
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  width: 100%;
  height: 100%;
}

.emotion-3:before {
  background-color: #f6f9fc;
  opacity: 0;
}

.emotion-3.fi-panel-expansion--open:before {
  opacity: 1;
  -webkit-transition: opacity 50ms cubic-bezier(0.28,0.84,0.42,1);
  transition: opacity 50ms cubic-bezier(0.28,0.84,0.42,1);
}

.emotion-3 .fi-panel-expansion-title {
  color: #282828;
  position: relative;
  display: block;
  width: 100%;
}

.emotion-3 .fi-panel-expansion-title:focus {
  outline: 0;
  position: relative;
}

.emotion-3 .fi-panel-expansion-title:focus:after {
  content: '';
  position: absolute;
  top: -4px;
  right: -4px;
  bottom: -4px;
  left: -4px;
  border-radius: 4px;
  background-color: transparent;
  border: 1px solid #ea7125;
  box-sizing: border-box;
  box-shadow: 0 0 10px 0 #ea7125;
  z-index: 9999;
}

.emotion-3 .fi-panel-expansion-title:focus:not(:focus-visible):after {
  content: none;
}

.emotion-3 .fi-panel-expansion-title--no-tag {
  font-family: 'Source Sans Pro','Helvetica Neue',Arial;
  font-size: 18px;
  line-height: 18px;
  font-weight: 400;
  line-height: 1em;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  white-space: nowrap;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-font-smoothing: antialiased;
  font-size: 14px;
  line-height: 14px;
  font-weight: 600;
  font-size: 18px;
  line-height: 18px;
  padding: 20px 60px 20px 20px;
  color: #2A6EBB;
}

.emotion-3 .fi-panel-expansion-title-icon {
  position: absolute;
  top: 0;
  right: 0;
  margin: 20px;
}

.emotion-3 .fi-panel-expansion-title--open .fi-panel-expansion-title-icon,
.emotion-3 .fi-panel-expansion-title-icon--open {
  -webkit-transform: rotate(-180deg);
  -ms-transform: rotate(-180deg);
  transform: rotate(-180deg);
}

.emotion-3 > .fi-panel-expansion-content {
  position: relative;
  display: block;
  height: 0;
  overflow: hidden;
  -webkit-transition: all 50ms cubic-bezier(0.28,0.84,0.42,1);
  transition: all 50ms cubic-bezier(0.28,0.84,0.42,1);
  padding: 0 20px;
}

.emotion-3 > .fi-panel-expansion-content.fi-panel-expansion-content--open {
  height: auto;
  overflow: visible;
  padding: 0 20px 20px 20px;
}

.emotion-1 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  -webkit-appearance: button;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font-size: 100%;
  font: inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: auto;
  display: inline-block;
  max-width: 100%;
  pointer: cursor;
}

.emotion-1:-moz-focusring {
  outline: 1px dotted ButtonText;
}

.emotion-1::-moz-focus-inner {
  border-style: none;
  padding: 0;
}

.emotion-1::-webkit-file-upload-button {
  -webkit-appearance: button;
  font: inherit;
}

.emotion-1::-webkit-inner-spin-button {
  height: auto;
}

.emotion-1::-webkit-outer-spin-button {
  height: auto;
}

.emotion-1,
.emotion-1 * {
  cursor: pointer;
}

.emotion-1:hover,
.emotion-1:active,
.emotion-1:focus,
.emotion-1:focus-within,
.emotion-1:visited {
  cursor: pointer;
}

.emotion-0 {
  display: inline-block;
  vertical-align: text-bottom;
}

.emotion-2 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font-size: 100%;
  font: inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: auto;
  display: block;
  max-width: 100%;
  display: block;
  width: 100%;
  max-width: 100%;
  display: none;
}

<div
  class="panel-expansion-test fi-panel-expansion emotion-3"
>
  <button
    aria-disabled="false"
    aria-expanded="false"
    class="fi-panel-expansion-title fi-panel-expansion-title--no-tag emotion-1"
    data-testid="panel-expansion-title"
    tabindex="0"
    type="button"
  >
    Test expansion
    <svg
      aria-hidden="true"
      class="fi-panel-expansion-title-icon emotion-0"
      focusable="false"
      height="1em"
      style="fill: #2A6EBB;"
      viewBox="0 0 24 24"
      width="1em"
    >
      <path
        d="M12 18c-.28 0-.548-.118-.737-.324l-8-9a1 1 0 0 1 1.474-1.352L12 15.52l7.263-8.196a1 1 0 1 1 1.474 1.352l-8 9A1.002 1.002 0 0 1 12 18"
      />
    </svg>
  </button>
  <div
    aria-hidden="true"
    class="fi-panel-expansion-content fi-panel emotion-2"
  >
    Test expansion content
  </div>
</div>
`;
